### ISSUE TITLE
microphone isn't working.

### ISSUE THREAD

**Issue by Legendary-Candice-Joe** [Feodor2/Mypal68#114]
When ever I use discord or a mic test, it shows the icon but no audio goes through.

I've actually had this issue on firefox its self so I'm not entirely blaming mypal for this one. But if I try to google a fix I usually just get articles with something really basic like "hAvE yOU TrIEd UNplUggInG iT aNd RePLuggIng IT iN???". Could be the Realtek audio drivers? since a lot of people don't seem to have this problem.

I've tinkered with audio settings and mic perms but nothing helped. In my case it mostly doesn't matter since I only needed it to work once but it would be nice for it to work.

Also I tried mypal 68 on another computer running windows xp and it had the exact same issue, that computer also had a Realtek driver so I guess that is the problem but I want to know if its possible to solve. Or if anyone else has had this issue.

**Comment by Feodor2** [comment]
check please media.peerconections must be true

**Comment by Feodor2** [comment]
Problem is not with the setting and this is winxp only bug.

**Comment by gdl-blue** [comment]
> Problem is not with the setting and this is winxp only bug.

Yes. I tried on Windows 7 but it worked fine. only on xp. Are you planning to fix it? (Note that in Centaury, microphone worked properly)

**Comment by pbcairo** [comment]
mic does work on winxp 32-bit Centaury 0.17.0

**Comment by Legendary-Candice-Joe** [comment]
I forgot to comment on this issue but it seems like even on mypal 68 13 (latest build) it still doesn't seem to work. And yeah I already have set the peerconnections to true.

**Comment by blue-dragon-cairo** [comment]
another Discord voice issue - an error in the console:
```[AudioActionCreators] unknown getUserMedia error: TypeError: e.getAudioContext().audioWorklet is undefined```

in about:config I set `dom.audioworklet.enabled` to true & browser crashes when using Discord voice

**Comment by Legendary-Candice-Joe** [comment]
That's weird.
And it's weirder that this is a windows XP only bug.

**Comment by blue-dragon-cairo** [comment]
ah yes I also tried Mypal on Windows 7 and mic input worked. 

Centaury was able to get mic working on XP so it wouldn't be impossible to fix it on Mypal-68

**Comment by Legendary-Candice-Joe** [comment]
Welp, new release.
Upload issue has indeed been fixed but the mic hasn't :(

I almost wonder if this is a driver problem since it didn't work on both Windows XP or Vista (dual booting on the same machine). I have a spare PCIe sound card that isn't a Realtek one. Before I was just using on board audio

Though I'm not even sure if it's worth testing or not. That being said I have not tested MyPal 68 in Windows 7, only XP and Vista so...
I'm sure this issue wouldn't be super hard to fix if it apparently works on at least Windows 7, but I don't actually know how to compile Firefox cause I've never done that before.

**Comment by Legendary-Candice-Joe** [comment]
I tried a USB audio mic and audacity can read the input but still mypal can't.
I'll try windows vista later and see if it works there.

**Comment by Feodor2** [comment]
The code shim for sound recording on winxp has been removed long ago. I am not going restore it because its big and not aligned with the new code. I shall do new shim, winmm sound recording seems still present on the webtrc, the goal is to connect it with the cubeb winmm part properly, not a priority though.

**Comment by dainiusb** [comment]
I have recording working.

[mypal68_winmm_mic_capture.patch](https://github.com/user-attachments/files/30993050/mypal68_winmm_mic_capture.patch)

**Comment by Feodor2** [comment]
Great, thanks a lot

**Comment by Feodor2** [comment]
How did you check? I just tried to record sound on xp and failed.

**Comment by dainiusb** [comment]
First of all I didn't compile, I hacked the dll directly so had to write the patch afterwards, mistake is possible of course. I can send you the modified xul.dll that works for me for test.
I had webrtc enabled. But probably not needed for simple recording.

**Comment by dainiusb** [comment]
I think one of the test sites I used was https://mictests.com/, which I saw didn't react like it should (noise displayed instead of bars), but it did record fine.
Another was this, I think:
https://webcammictest.com/mic/
And also my own test page.
Worked on wine as well.

You're testing on real hardware?

**Comment by dainiusb** [comment]
I see the patch does not exactly represent the modifications done to the dll.

Try this one:

[mypal68_winmm_mic_capture2.patch](https://github.com/user-attachments/files/31331216/mypal68_winmm_mic_capture2.patch)

If not working then this one:

[mypal68-winmm-capture-reconstruction.patch](https://github.com/user-attachments/files/31331226/mypal68-winmm-capture-reconstruction.patch)

Thanks for your feedback.

**Comment by Feodor2** [comment]
Great! The last patch actually fixed it.
